Firestore/core/src/firebase/firestore/model/field_value.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_VALUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_VALUE_H_
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * Abstract base class representing an immutable data value as stored in
+ * Firestore. FieldValue represents all the different kinds of values
+ * that can be stored in fields in a document.
+ */
+class FieldValue {
+ public:
+
+  /** The order of types in Firestore; this order is defined by the backend. */
+  typedef enum {
+    TypeOrderNull,
+    TypeOrderBoolean,
+    TypeOrderNumber,
+    TypeOrderTimestamp,
+    TypeOrderString,
+    TypeOrderBlob,
+    TypeOrderReference,
+    TypeOrderGeoPoint,
+    TypeOrderArray,
+    TypeOrderObject,
+  } TypeOrder;
+
+  /**
+   * All the different kinds of values that can be stored in fields in
+   * a document.
+   */
+  typedef enum {
```

Comment:
Similar feedback here: `enum class Type`. Also Type is the principal concept here and TypeOrder is secondary.

You may even be able to make `TypeOrder` a private enum.